src/amqp-client/channel.cr

```diff
@@ -255,25 +254,44 @@ class AMQP::Client
         end
       end
       if @confirm_mode
-        @confirm_id = @confirm_id &+ 1_u64
+        @confirm_lock.synchronize do
+          msgid = @confirm_id &+= 1_u64
+          @unconfirmed_publishes.push msgid
+          msgid
+        end
       else
         0_u64
       end
     end
 
     def basic_publish_confirm(msg, exchange, routing_key = "", mandatory = false, immediate = false, props properties = Properties.new) : Bool
       confirm_select
-      msgid = basic_publish(msg, exchange, routing_key, mandatory, immediate, properties)
-      wait_for_confirm(msgid)
+      basic_publish(msg, exchange, routing_key, mandatory, immediate, properties)
+      wait_for_confirms
     end
 
     def basic_publish_confirm(io : IO, bytesize : Int, exchange : String, routing_key = "", mandatory = false, immediate = false, props properties = Properties.new) : Bool
       confirm_select
-      msgid = basic_publish(io, bytesize, exchange, routing_key, mandatory, immediate, properties)
-      wait_for_confirm(msgid)
+      basic_publish(io, bytesize, exchange, routing_key, mandatory, immediate, properties)
+      wait_for_confirms
+    end
+
+    # Returns when there are no unconfirmed publishes on the channel
+    # Raises if there was any negative acknowledgements
+    def wait_for_confirms : Bool
+      ok = @unconfirmed_empty.receive
```

Comment:
wonder if we should check if the dequeue is empty first? could get stuck here if the user hasn't called this method while the ack comes back from the server, say: 

```crystal
ch.confirm_select
ch.basic_publish "hello world"
sleep 1
ch.wait_for_confirms # will hang forever
```

or just
```crystal
ch = connection.channel
ch.wait_for_confirms # will hang forever
```